Rework handling of compositions and projects, and composer windows

nyalldawson opened the pull request:
This PR implements a major rework to how compositions are managed. 

Highlights are:

- A new QgsLayoutManager class which is used for storage and serializing/deserializing of compositions. A QgsLayoutManager is attached to QgsProject. This allows core code to access the compositions attached to a project. All handling of compositions has been moved from app to core, making it easy for server (and other non-app based projects. QField/roam?) to access project compositions without resorting to fragile xml parsing.
- Composer windows are created on demand and destroyed on close. This avoids the heavy requirement of creating all composer windows and widgets for every single composer item when loading projects. It's the main reason loading projects with compositions is so slow currently. Now, composer windows are created only when the window is opened and destroyed when the window is closed. Composer item configuration widgets are also only created on demand (when an item is selected) instead of precreating all widgets.
- A side benefit of destroying composer windows on close is that we now no longer suffer from a Qt bug which blocked use of floating docks in composer windows. Now you can float the item/composition/etc panels! Much nicer for full screen working with compositions on multi-monitor setups.
- Reworked API for working with composers through iface. Since composer windows don't exist unless they are currently open, I've redone all the iface methods for interacting with composers to make it clear that they only apply to open windows (code which works with compositions (not composers) should be updated to use the layoutManager() attached to the active project instead, since compositions in the manager always exist). Additionally, I've added a simple composer interface which provides a stable API for plugins and scripts to work with composer windows. Very basic for now, but in future we can extend with more hooks to allow plugins to interact with composer windows.
- Lots of code cleanups and removal

Sponsored by the Swiss QGIS User Group!

nyalldawson committed:
[composer] Attach compositions to QgsProject

This adds a new QgsLayoutManager class which is used for
storage and serializing/deserializing of compositions.

A QgsLayoutManager is attached to QgsProject. This allows
core code to access the compositions attached to a project.
The intention is to move all handling of compositions from
app to core, making it easy for server to access project
compositions without resorting to fragile xml parsing.

nyalldawson committed:
Followup setting composition name

nyalldawson committed:
Use layout manager when testing for unique composer names

nyalldawson committed:
[composer] Composer windows are created on demand and destroyed on close

Avoids having to create a window and all widgets for every
composition when loading a project. Speeds up loading
projects.

Fix #15193

nyalldawson committed:
Remove more unneeded composer window code

nyalldawson committed:
[composer] Allow floating docks

Now that composer windows are deleted on close, we no longer
suffer from the Qt bug which keeps floating docks open when
the window is hidden

nyalldawson committed:
Port composer template saving and duplicating to layout manager

nyalldawson committed:
Remove more ununused composer code, fix updating existing
compositions when settings change

nyalldawson committed:
Extend layout manager tests

nyalldawson committed:
Restore opening existing composers

nyalldawson committed:
Better method to autogenerate composer names

nyalldawson committed:
Add a basic interface for composer windows

Not much there yet, but it adds an interface with stable API
which can be used by plugins and scripts to interact with
composer windows.

Could be expanded in future to add hooks for toolbars, menus,
etc, allowing plugins to extend composer functionality.

nyalldawson committed:
[composer] Rework interface composer methods

Since composer windows are now only created on demand, and
destroyed when the window is closed, the old interface methods
no longer apply. The new interface methods openComposers(),
openComposer(), closeComposer(), composerOpened(),
composerWillBeClosed() and composerClosed() are similar, but
only apply to composer windows when they are exist (i.e.
are currently open).

To access all compositions from a project, the project's
layoutManager() should be used instead.

Additionally, the new interface methods work with
QgsComposerInterface objects instead of QgsComposerView
objects. This should allow future hooks for plugins to interact
with open composer windows in a more useful way.

nyalldawson committed:
[composer] Port composer manager to layout manager

Also switch to using a proper model to show/edit compositions
in the dialog.

nyalldawson committed:
Remove another use of old iface methods

nyalldawson committed:
Rework atlas layer action to handle temporary composer windows

nyalldawson committed:
Only create composer item config widgets on demand

Instead of creating them for all items when composer window
is opened, instead just create and destroy them when
required. None are too heavy to have a noticable delay
when selecting items, but in contrast keeping them around
forever is wasteful on memory.

Also clean up a lot of duplicate composer/composition
code and remove unnecessary signals from api (use a single
itemAdded signal instead of multiple signals for every
item type)

nyalldawson committed:
Fix crash when reopening composers

wonder-sk commented on python/core/composer/qgslayoutmanager.sip:
```diff
@@ -0,0 +1,34 @@
+%End
+  public:
+
+    explicit QgsLayoutManager( QgsProject* project = 0 );
```
Probably /TransferThis/ annotation should be here?

nyalldawson replied:
Fixed

wonder-sk commented on python/core/composer/qgslayoutmanager.sip:
```diff
@@ -0,0 +1,34 @@
+    bool readXml( const QDomElement &element, const QDomDocument &doc );
+    QDomElement writeXml( QDomDocument &doc ) const;
+    bool saveAsTemplate( const QString &name, QDomDocument &doc ) const;
+    QgsComposition *duplicateComposition( const QString &name, const QString &newName );
```
/Factory/ annotation for duplicateComposition()?

nyalldawson replied:
No, it's not required - the new composition is automatically added to the manager and ownership remains with the manager.

wonder-sk reviewed (commented):
This looks great! I haven't looked in detail at all the changes but the architecture changes are exactly what we needed for a long time...

m-kuhn commented:
Nice one!

I wouldn't mind having a new composer module next to core/gui/analysis, after all for custom android apps it will probably not always be required and then the app size could be reduced by stripping the library out, but I'm not sure how much it's interconnected with other core functionality. Most important is: it's out of gui.

timlinux commented:
Nice stuff @nyalldawson ! I added your nice commit message to a place holder entry in changelog.

http://changelog.qgis.org/en/qgis/version/3.0.0/#map-composer-overhaul

Are there any gotchas plugin / python users will have to look out for if they are constructing composer templates programmatically and need to port their code?

nyalldawson commented:
> Are there any gotchas plugin / python users will have to look out for if they are constructing composer templates programmatically and need to port their code?

It depends. It's all documented in the api break dox, but if you're just using creating and working with compositions outside of projects, and not using the iface methods, then nothing's changed (yet). It's only if you're interacting with open composers within the iface.

nyalldawson committed:
Don't use combobox of existing composer names in new composer title dialog

It's odd UX - selecting any of these results in an error message,
so giving them as choices is misleading.

Fix #12673 (indirectly)

nyalldawson committed:
Add missing TransferThis annotation

nyalldawson committed:
Fix restoration of compositions from legacy projects

nyalldawson committed:
Add/refine test masks for server test

nyalldawson committed:
Fix loading composition triggers render of all composer maps

Also cleanup some more unused API methods

nyalldawson committed:
Remove hack to get composer maps to follow canvas

nyalldawson committed:
Avoid multiple map/legend renders while loading compositions

nyalldawson committed:
Fix Windows build

nyalldawson committed:
Fix missing initial render of composer maps

nyalldawson committed:
Better handling of composer map forced refreshes

nyalldawson committed:
[composer] Don't load HTML for labels on project load

Instead defer to composer load. Also make sure html is reloaded
when composition is force refreshed.

nirvn commented:
@nyalldawson , working flawlessly on my machine. Project loading is now blazingly fast, wouhou!

NathanW2 commented:
Nice work!

On Tue, Mar 21, 2017 at 1:25 PM, Nyall Dawson <notifications@github.com>
wrote:

> Merged #4279 <https://github.com/qgis/QGIS/pull/4279>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4279#event-1007876433>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3IOYd2Yo8dtA2gkOiDZB3JTUoMxOks5rn0M4gaJpZM4MhYkp>
> .
>
